### PR TITLE
Bugfix 1216: Use revision marker to store stream offset in segment attribute.

### LINKED ISSUE #1216: Store compaction point in segment metadata
**Problem description**
More data than is necessary may be re-read to create a new State synchronizer impl. 

**Problem location**
StateSynchronizerImpl

**Suggestions for an improvement**
The offset of the last compaction can be stored in the segment metadata so reading can skip to exactly that point.

### PR DESCRIPTION
**Pull request by tkaitchuck**
**Change log description**
* Add marker support in RevisionedStreamClient
* Have StateSynchronizer use the marker to store the last compaction point.

**Purpose of the change**
Enables marking a position in a segment using a SegmentAttribute.
Which in turn is used to prevent over-reading when creating a new StateSynchronizer.
Fixes #1216 

**What the code does**
Connects to the new SegmentAttribute code and exposes a feature taking advantage of it on the RevisionedStreamClient. It sets up an internal class SegmentMetadataClient for managing these updates and enum to track all of the SegmentAttributes that are used in the client.

**How to verify it**
New tests have been added to cover the new code paths. All existing tests pass.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java [hunk 1]
```diff
@@ -205,6 +187,9 @@ public ReadFuture read(long offset, int length) {
         getConnection().thenAccept((ClientConnection c) -> {
             log.debug("Sending read request {}", read);
             c.sendAsync(read.request);
+        }).exceptionally(e -> {
+            read.completeExceptionally(new RuntimeException(e));
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java]: Isn't the convention to wrap it in a `CompletionException`?

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.segment.impl;
+
+/**
+ * A client for looking at an editing the metadata related to a specific segment.
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java]: "at the metadata"?

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ */
+public interface SegmentMetadataClient {
+    
+    /**
+     * Returns the length of the current segment. i.e. the total length of all data written to the segment.
+     *
+     * @return The length of the current segment.
+     */
+    abstract long fetchCurrentStreamLength();
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java]: why `fetch` and not `get`? You use `get` below for attributes.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientFactory.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+package io.pravega.client.segment.impl;
+
+/**
+ * Creates {@link SegmentMetadataClient} for metadata operations on existing segments.
+ */
+public interface SegmentMetadataClientFactory {
+    /**
+     * Opens an existing segment for metadata operations. This operation will fail if the
+     * segment does not exist.
+     * This operation may be called multiple times on the same stream from the
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientFactory.java]: stream -> segment

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+        @Override
+        public void processingFailure(Exception error) {
+            log.warn("Processing failure: ", error);
+            closeConnection(error);
+        }
+    }
+
+    private void closeConnection(Exception exceptionToInflightRequests) {
+        log.info("Closing connection with exception: {}", exceptionToInflightRequests.toString());
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java]: no need to call `toString`

Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java]: same below

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            infoRequests.put(requestId, result);
+        }
+        getConnection().thenAccept(c -> {
+            try {
+                log.debug("Getting segment info for segment: {}", segmentId);
+                c.send(new WireCommands.GetStreamSegmentInfo(requestId, segmentId.getScopedName()));
+            } catch (ConnectionFailedException e) {
+                closeConnection(e);
+            }
+        }).exceptionally(e -> {
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java]: You can use `FutureHelpers.exceptionListener` to simplify this syntax (i.e., not having to add `return null` at the end).

File: client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java (hunk 1 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java]: Done

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java (hunk 2 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java]: fixed.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientFactory.java (hunk 4 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientFactory.java]: done

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java (hunk 5 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java]: switched to getMessage()

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java (hunk 6 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java]: done

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+        }
+        FutureHelpers.exceptionListener(getConnection().thenAccept(c -> {
+            try {
+                log.debug("Getting segment info for segment: {}", segmentId);
+                c.send(new WireCommands.GetStreamSegmentInfo(requestId, segmentId.getScopedName()));
+            } catch (ConnectionFailedException e) {
+                closeConnection(e);
+            }
+        }), e -> {
+            result.completeExceptionally(new RuntimeException(e));
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java]: This is still RuntimeException. Same below (6 occurences)

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        });
+        return result;
+    }
+    
+    @Override
+    public long fetchCurrentStreamLength() {
+        return RETRY_SCHEDULE.retryingOn(ConnectionFailedException.class)
+                             .throwingOn(InvalidStreamException.class)
+                             .run(() -> {
+                                 return FutureHelpers.getThowingException(getSegmentInfo()).getSegmentLength();
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java]: getThrowingException

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java (hunk 3 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java]: I changed both to fetch to make it clear that IO is being performed.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java (hunk 7 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java]: I found and fixed 3 more in this file. Are there others?

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java (hunk 8 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java]: fixed.

**Pull request merged**